### DISCUSSION TITLE
extra-substituters are not used

### DISCUSSION THREAD

**Discussion by tianze0926** [ryan4yin/nixos-and-flakes-book#62] [Q&A] [answered]
I have been using NixOS with flakes for a few months, thanks to this amazing book. But recently I noticed that the cachix substituters I put in `nixConfig.extra-substituters` were totally useless ([here](https://github.com/tianze0926/config/blob/f54de100b0b4c4c532d15be5add6311d002ed708/flake.nix#L10C27-L10C27) is my configuration). The binary cache is not used and Nix starts the building process locallly. Since I use http proxy, I can see the network traffic of Nix, and Nix didn't even access the binary cache such as https://nix-community.cachix.org.

But if I move the substituters to `nixConfig.substituters`, the binary cache would be used. I am very confused about this. Could anyone help me about it? Thanks~

**Comment by ryan4yin** [comment, marked as answer]
Good question! I haven't paid much attention to whether this parameter is valid before.
I spent some time studying this problem. Here is a detailed explanation of the background and reasons for this problem.

## `nixConfig` in flake.nix

According to [Flake format - Nix Manual](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake.html?highlight=nixconfig#flake-format):

> `nixConfig`: a set of `nix.conf` options to be set when evaluating any part of a flake. In the interests of security, only a small set of set of options is allowed to be set without confirmation so long as [accept-flake-config](https://nixos.org/manual/nix/stable/command-ref/conf-file#conf-accept-flake-config) is not enabled in the global configuration:

So options in `nixConfig` will **only take effect in the flake itself**, and to to specify additional [substituters](https://nixos.org/manual/nix/stable/command-ref/conf-file#conf-substituters) by `nixConfig.substituters` or `nixConfig.extra-substituters`, the current user must be setted in [`trusted-users` in the system-level `/etc/nix/nix.conf`](https://nixos.org/manual/nix/stable/command-ref/conf-file#conf-trusted-users).

## `nix.settings` and `/etc/nix/nix.conf`

To customize the system-level `/etc/nix/nix.conf` declaratively, we should add something like this in a nixos module:

```nix
{

  nix.settings = {
    # enable flakes globally
    experimental-features = ["nix-command" "flakes"];
    # customize the system-level substituers
    substituters = [
      "https://mirrors.cernet.edu.cn/nix-channels/store"
      "https://cache.nixos.org/"
    ];

    # given the users in this list the right to specify additional substituters via:
    #    1. `nixConfig.substituers` in `flake.nix`
    #    2. command line args `--options substituers http://xxx`
    trusted-users = ["user1" "user2"];
  };
}
```

## `extra-` prefix in `nix.conf`

Here is docs about `extra-` prefix in [`nix.conf` - file format - nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html?highlight=extra#file-format)

> A configuration setting usually overrides any previous value. However, for settings that take a list of items, you can prefix the name of the setting by `extra-` to append to the previous value.

So the content in `nixConfig.extra-substituers` or `nix.setting.extra-substituers` should be appended to the `substituers` option in the current scope.

But why did `nixConfig.extra-substituers` is ignored by nix? I found several related issues:

* https://github.com/NixOS/nix/issues/9487
* https://github.com/NixOS/nix/issues/7560
* https://github.com/NixOS/nix/issues/6170

As the issues showed here, the `extra-substituers` options may be ignored in many cases.

I would recommend using `nix.settings.substituers` to set the system-level generic cache servers, and `nixConfig.extra-substituers` for the flake-only cache servers.

But if you encounter any problem with `nixConfig.extra-substituers` being ignored, you can specify all the cache servers you need to solve the problem directly through `nixConfig.substituers`.

**Comment by ryan4yin** [comment]
Some new content has been added to this: https://github.com/ryan4yin/nixos-and-flakes-book/commit/8d3ec992fe686fcf5cfe1418379e40cd1e4564b1
